doc/api/async_hooks.md

```diff
@@ -0,0 +1,677 @@
+
+Registers functions to be called for different lifetime events of each async
+operation.
+The callbacks `init()`/`before()`/`after()`/`destroy()` are registered via an
+`AsyncHooks` instance and fire during the respective asynchronous events in the
+lifetime of the event loop. The focal point of these calls centers around the
+lifetime of the `AsyncWrap` C++ class. These callbacks will also be called to
+emulate the lifetime of handles and requests that do not fit this model. For
+example, `HTTPParser` instances are recycled to improve performance. Therefore the
+`destroy()` callback is called manually after a connection is done using
+it, just before it is placed back into the unused resource pool.
+
+All callbacks are optional. So, for example, if only resource cleanup needs to
+be tracked then only the `destroy()` callback needs to be passed. The
+specifics of all functions that can be passed to `callbacks` is in the section
+`Hook Callbacks`.
+
+##### Error Handling
+
+If any `AsyncHook` callbacks throw, the application will
+print the stack trace and exit. The exit path does follow that of any uncaught
+exception. However `'exit'` callbacks will still fire unless the application
+is run with `--abort-on-uncaught-exception`, in which case a stack trace will
+be printed and the application exits, leaving a core file.
+
+The reason for this error handling behavior is that these callbacks are running
+at potentially volatile points in an object's lifetime. For example, during
+class construction and destruction. Because of this, it is deemed necessary to
+bring down the process quickly in order to prevent an unintentional abort in the
+future. This is subject to change in the future if a comprehensive analysis is
+performed to ensure an exception can follow the normal control flow without
+unintentional side effects.
+
+
+#### `asyncHook.enable()`
+
+* Returns {AsyncHook} A reference to `asyncHook`.
+
+Enable the callbacks for a given `AsyncHook` instance. 
+
+```js
+const async_hooks = require('async_hooks');
+
+const hook = async_hooks.createHook(callbacks).enable();
+```
+
+#### `asyncHook.disable()`
+
+* Returns {AsyncHook} A reference to `asyncHook`.
+
+Disable the callbacks for a given `AsyncHook` instance from the global pool of
+hooks to be executed. Once a hook has been disabled it will not fire again
+until enabled.
+
+For API consistency `disable()` also returns the `AsyncHook` instance.
+
+#### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas: instantiation, before/after the callback is called, and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+##### `init(id, type, triggerId, resource)`
+
+* `id` {number} a unique id for the async resource
+* `type` {string} the type of the async resource
+* `triggerId` {number} the unique id of the async resource in whose
```

Comment:
I don't have strong feelings, but just to be clear, this is the first version of the docs I've read, so if it appears to me that what is being described is a parent relationship, and I'm having to guess that what trigger means is "parent", maybe the old name made more sense? If there was some problems with that name, trigger is fine, but it has to be described, its not obvious.